runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,350 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+local namespace = api.nvim_create_namespace('vim_lsp_document_highlight')
+local augroup = api.nvim_create_augroup('vim_lsp_document_highlight', {})
+
+local globalstate = {
+  enabled = false,
+}
+
+---Buffer-local state for document highlights
+---@class (private) vim.lsp.document_highlight.BufState
+---
+---Whether document highlights are enabled for this buffer,
+---`nil` indicates following the global state.
+---@field enabled? boolean
+---
+---Each data change generates a unique version,
+---not garanteed, numbers may be reused over time.
+---@field version integer
+---
+---Latest `version` with data applied.
+---@field applied? integer
+---
+---Each client attached to this buffer must exists.
+---
+---Index in the form of client_id -> (row -> highlights)
+---@field client_hilights table<integer, table<integer, lsp.DocumentHighlight[]?>?>
+
+---Each buffer attached by at least one supported LSP server must exists,
+---otherwise it should not exists or be cleaned up.
+---
+---Index in the form of bufnr -> bufstate
+---@type table<integer, vim.lsp.document_highlight.BufState?>
+local bufstates = {}
+for _, client in ipairs(vim.lsp.get_clients()) do
+  for _, bufnr in ipairs(vim.lsp.get_buffers_by_client_id(client.id)) do
+    if client.supports_method(ms.textDocument_documentHighlight, { bufnr = bufnr }) then
+      local bufstate = bufstates[bufnr] or {}
+      local client_hilights = bufstate.client_hilights or {}
+
+      if not client_hilights[client.id] then
+        client_hilights[client.id] = {}
+      end
+
+      bufstate.client_hilights = client_hilights
+      bufstates[bufnr] = bufstate
+    end
+  end
+end
+api.nvim_create_autocmd('LspAttach', {
```

Comment:
I don't think these should be created at `require` time. These should either be placed somewhere in `plugin`, or setup via some other function. Perhaps `lsp.start()`?

Reply:
Makes sense, but this part is likely temporary. I will add a dedicated PR for managing bufstate after a while(for this module and `inaly_hint.lua`), and the logic is based on this part, so these autocmds will be removed from this module regardless. And `inlay_hint.lua` has a similar logic, so maybe in follow-up PRs? Is there any vulnerability here?